FastLED no longer compiles against latest Adafruit nRF52 board definitions (gcc 7-2017q4) due to invalid C++ code in FastLED

In v0.11.0, Adafruit switched the version of gcc that they are using for the nRF52 boards

See this commit: https://github.com/adafruit/Adafruit_nRF52_Arduino/commit/bbe64fb9bb2ee2d8fbe6a7a11b8ae817379b7938#diff-d638a47d1ca0e2a5585fb34c7d7fb0f9

Along with their package.json (https://adafruit.github.io/arduino-board-index/package_adafruit_index.json), comparing lines 3264 and 3590.

The effect of this change is to move nRF52 build toolchain from gcc 5.2 to gcc 7.

One of the changes between these two versions of gcc was to fix a gcc bug that was allowing reinterpret_casts to happen inside constexprs (according to the spec this is not valid).  The compiler change is now causing the following compilation error when compiling for any of the nRF52 boards (nRF52832 & nRF52840):

```
In file included from /Users/user/Library/Arduino15/packages/adafruit/hardware/nrf52/0.11.1/cores/nRF5/nordic/nrfx/mdk/nrf.h:93:0,
                 from /Users/user/Library/Arduino15/packages/adafruit/hardware/nrf52/0.11.1/cores/nRF5/WVariant.h:23,
                 from /Users/user/Library/Arduino15/packages/adafruit/hardware/nrf52/0.11.1/cores/nRF5/Arduino.h:40,
                 from /var/folders/c2/wdh0gcts5qn6rf1npbsnxf180000gn/T/arduino_build_147358/sketch/sketch_jul17a.ino.cpp:1:
/Users/user/Arduino/libraries/FastLED/platforms/arm/nrf52/fastpin_arm_nrf52.h: In static member function 'static constexpr NRF_GPIO_Type* __generated_struct_NRF_P0::r()':
/Users/user/Library/Arduino15/packages/adafruit/hardware/nrf52/0.11.1/cores/nRF5/nordic/nrfx/mdk/nrf52840.h:2826:38: error: reinterpret_cast from integer to pointer
 #define NRF_P0                      ((NRF_GPIO_Type*)          NRF_P0_BASE)
                                     ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/user/Arduino/libraries/FastLED/platforms/arm/nrf52/fastpin_arm_nrf52.h:77:16: note: in expansion of macro 'NRF_P0'
         return NRF_P0;
                ^~~~~~
/Users/user/Arduino/libraries/FastLED/platforms/arm/nrf52/fastpin_arm_nrf52.h: In static member function 'static constexpr NRF_GPIO_Type* __generated_struct_NRF_P1::r()':
/Users/user/Library/Arduino15/packages/adafruit/hardware/nrf52/0.11.1/cores/nRF5/nordic/nrfx/mdk/nrf52840.h:2827:38: error: reinterpret_cast from integer to pointer
 #define NRF_P1                      ((NRF_GPIO_Type*)          NRF_P1_BASE)
                                     ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/user/Arduino/libraries/FastLED/platforms/arm/nrf52/fastpin_arm_nrf52.h:82:16: note: in expansion of macro 'NRF_P1'
         return NRF_P1;
                ^~~~~~
```

The problematic code is here in fastpin_arm_nrf52.h:

```
struct __generated_struct_NRF_P0 {
      FASTLED_NRF52_INLINE_ATTRIBUTE constexpr static NRF_GPIO_Type * r() {
          return NRF_P0;
      }
  };
```

NRF_P0 is a macro which expands to 

```
#define NRF_P0                      ((NRF_GPIO_Type*)          NRF_P0_BASE)
```

and NRF_P0_BASE is just the memory address of P0

```
#define NRF_P0_BASE                 0x50000000UL
```

Unfortunately my C++ template metaprogramming skills are not good enough to try to fix this and submit a PR.

--- a/platforms/arm/nrf52/fastpin_arm_nrf52.h
+++ b/platforms/arm/nrf52/fastpin_arm_nrf52.h
@@ -73,13 +73,13 @@
 
 // manually define two structures, to avoid fighting with preprocessor macros
 struct __generated_struct_NRF_P0 {
-    FASTLED_NRF52_INLINE_ATTRIBUTE constexpr static NRF_GPIO_Type * r() {
-        return NRF_P0;
+    FASTLED_NRF52_INLINE_ATTRIBUTE constexpr static uintptr_t r() {
+        return NRF_P0_BASE;
     }
 };
 struct __generated_struct_NRF_P1 {
-    FASTLED_NRF52_INLINE_ATTRIBUTE constexpr static NRF_GPIO_Type * r() {
-        return NRF_P1;
+    FASTLED_NRF52_INLINE_ATTRIBUTE constexpr static uintptr_t r() {
+        return NRF_P1_BASE;
     }
 };
 
@@ -112,19 +112,19 @@ public:
         NRF_GPIO_PIN_NOSENSE            // pin sense level disabled
         );
   }
-  FASTLED_NRF52_INLINE_ATTRIBUTE static void       hi()        { _PORT::r()->OUTSET = _MASK;            } // sets _MASK in the SET   OUTPUT register (output set high)
-  FASTLED_NRF52_INLINE_ATTRIBUTE static void       lo()        { _PORT::r()->OUTCLR = _MASK;            } // sets _MASK in the CLEAR OUTPUT register (output set low)
-  FASTLED_NRF52_INLINE_ATTRIBUTE static void       toggle()    { _PORT::r()->OUT ^= _MASK;              } // toggles _MASK bits in the OUTPUT GPIO port directly
+  FASTLED_NRF52_INLINE_ATTRIBUTE static void       hi()        { (reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUTSET = _MASK;            } // sets _MASK in the SET   OUTPUT register (output set high)
+  FASTLED_NRF52_INLINE_ATTRIBUTE static void       lo()        { (reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUTCLR = _MASK;            } // sets _MASK in the CLEAR OUTPUT register (output set low)
+  FASTLED_NRF52_INLINE_ATTRIBUTE static void       toggle()    { (reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUT ^= _MASK;              } // toggles _MASK bits in the OUTPUT GPIO port directly
   FASTLED_NRF52_INLINE_ATTRIBUTE static void       strobe()    { toggle();     toggle();                } // BUGBUG -- Is this used by FastLED?  Without knowing (for example) SPI Speed?
-  FASTLED_NRF52_INLINE_ATTRIBUTE static port_t     hival()     { return _PORT::r()->OUT | _MASK;        } // sets all _MASK bit(s) in the OUTPUT GPIO port to 1
-  FASTLED_NRF52_INLINE_ATTRIBUTE static port_t     loval()     { return _PORT::r()->OUT & ~_MASK;       } // sets all _MASK bit(s) in the OUTPUT GPIO port to 0
-  FASTLED_NRF52_INLINE_ATTRIBUTE static port_ptr_t port()      { return &(_PORT::r()->OUT);             } // gets raw pointer to OUTPUT          GPIO port
-  FASTLED_NRF52_INLINE_ATTRIBUTE static port_ptr_t cport()     { return &(_PORT::r()->OUTCLR);          } // gets raw pointer to SET   DIRECTION GPIO port
-  FASTLED_NRF52_INLINE_ATTRIBUTE static port_ptr_t sport()     { return &(_PORT::r()->OUTSET);          } // gets raw pointer to CLEAR DIRECTION GPIO port
+  FASTLED_NRF52_INLINE_ATTRIBUTE static port_t     hival()     { return (reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUT | _MASK;        } // sets all _MASK bit(s) in the OUTPUT GPIO port to 1
+  FASTLED_NRF52_INLINE_ATTRIBUTE static port_t     loval()     { return (reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUT & ~_MASK;       } // sets all _MASK bit(s) in the OUTPUT GPIO port to 0
+  FASTLED_NRF52_INLINE_ATTRIBUTE static port_ptr_t port()      { return &((reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUT);             } // gets raw pointer to OUTPUT          GPIO port
+  FASTLED_NRF52_INLINE_ATTRIBUTE static port_ptr_t cport()     { return &((reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUTCLR);          } // gets raw pointer to SET   DIRECTION GPIO port
+  FASTLED_NRF52_INLINE_ATTRIBUTE static port_ptr_t sport()     { return &((reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUTSET);          } // gets raw pointer to CLEAR DIRECTION GPIO port
   FASTLED_NRF52_INLINE_ATTRIBUTE static port_t     mask()      { return _MASK;                          } // gets the value of _MASK
   FASTLED_NRF52_INLINE_ATTRIBUTE static void hi (register port_ptr_t port) { hi();                      } // sets _MASK in the SET   OUTPUT register (output set high)
   FASTLED_NRF52_INLINE_ATTRIBUTE static void lo (register port_ptr_t port) { lo();                      } // sets _MASK in the CLEAR OUTPUT register (output set low)
-  FASTLED_NRF52_INLINE_ATTRIBUTE static void set(register port_t     val ) { _PORT::r()->OUT = val;     } // sets entire port's value (optimization used by FastLED)
+  FASTLED_NRF52_INLINE_ATTRIBUTE static void set(register port_t     val ) { (reinterpret_cast<NRF_GPIO_Type*>(_PORT::r()))->OUT = val;     } // sets entire port's value (optimization used by FastLED)
   FASTLED_NRF52_INLINE_ATTRIBUTE static void fastset(register port_ptr_t port, register port_t val) { *port = val; }
   constexpr                      static uint32_t   nrf_pin2() { return NRF_GPIO_PIN_MAP(_PORT_NUMBER, _PIN_NUMBER); }
   constexpr                      static bool       LowSpeedOnlyRecommended() {
@@ -269,7 +269,7 @@ public:
             );
     #endif
     #if !defined(_FASTLED_NRF52_LOW_SPEED_ONLY_BOARD_DETECT)
-        #warning "Unknown board / package, ... caller must pins support high-speed"
+        #warning "Unknown board / package, ... caller must determine pins that support high-speed"
         return false; // choosing default to be FALSE, to allow users to ATTEMPT to use high-speed on pins where support is not known
     #endif
   }

